Correctly support updates in EnsureTCPLoadBalancer for GCE.

a-robinson opened the pull request:
Previously we'd just tear everything down and recreate it, which makes
for a pretty bad experience because it causes downtime whenever the
service controller restarts and has to make sure everything is in the
desired state.

This adds more code than I'd prefer, but makes it much cleaner and more
organized than it was before, in my opinion (despite all the update-related
additions). I didn't bother parallelizing anything because it's complex enough
as it is, right now.

It's consistently passing the existing e2es and worked when I tested
manually, but this could definitely use additional e2e tests and/or some
serious refactoring to make real unit tests feasible. I'll follow up
with one or two e2e tests that make sense (updating an LB or killing the
controller manager, perhaps).

Fixes #13333

@thockin @cjcullen @davidopp

a-robinson committed:
Correctly support updates in EnsureTCPLoadBalancer for GCE.

Previously we'd just tear everything down and recreate it, which makes
for a pretty bad experience because it causes downtime whenever the
service controller restarts and has to make sure everything is in the
desired state.

This adds more code than I'd prefer, but makes it much cleaner and more
organized than it was before, in my opinion. I didn't bother
parallelizing anything because it's complex enough as it is, right now.

It's consistently passing the existing e2es and worked when I tested
manually, but this could definitely use additional e2e tests and/or some
serious refactoring to make real unit tests feasible. I'll follow up
with one or two e2e tests that make sense (updating an LB or killing the
controller manager, perhaps).

a-robinson commented:
Ping - @cjcullen or @brendandburns up for reviewing this? It needs to get into 1.1 to fix #13333

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -333,114 +309,150 @@ func isHTTPErrorCode(err error, code int) bool {
+	// can't delete a target pool that's currently in use by a forwarding rule.
+	// Thus, we have to tear down the forwarding rule if either it or the target
+	// pool needs to be updated.
+	if !fwdRuleExists && (fwdRuleNeedsUpdate || tpNeedsUpdate) {
```
Shouldn't this be `if fwdRuleExists && ...` ?

a-robinson replied:
Indeed it should, thanks. This is why an e2e test is needed :)

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -333,114 +309,150 @@ func isHTTPErrorCode(err error, code int) bool {
-		if !ownsAddress(loadBalancerIP, addresses.Items) {
-			return nil, fmt.Errorf("this gce project don't own the IP address: %s", loadBalancerIP.String())
+	}
+	if !tpExists && tpNeedsUpdate {
```
Same here. Shouldn't it be `if tpExists & ...` ?

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -333,114 +309,150 @@ func isHTTPErrorCode(err error, code int) bool {
+		return nil, err
+	}
+
+	// Now we get to some slightly more interesting logic.
```
I think it'd be clearer to structure this a little differently (because this logic is damn confusing).

I was thinking
```
if fwdRuleNeedsUpdate || tpNeedsUpdate {
	if fwdRuleExists {
		// delete forwarding rule.
	}
	if tpNeedsUpdate {
		if tpExists {
			// delete target pool.
		}
		// create target pool.
	}
	// create forwarding rule
}
```
I'm also fine if you think it's clearer as is.

a-robinson replied:
I played around with it, and it didn't really seem to make anything clearer, just much more heavily nested. The code was exactly the same, just nested more, which to me made it look even less understandable. Happy to chat about this offline if you'd like.

cjcullen replied:
I'll buy the excessive nesting argument... I still think it would make it clearer/harder to mess up, but it's not worth holding up this PR.

davidopp commented:
I don't know this code well enough to review it (hopefully some day...), but thanks a lot for doing this!

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -333,114 +309,150 @@ func isHTTPErrorCode(err error, code int) bool {
+	return status, nil
+}
+
+func (gce *GCECloud) forwardingRuleNeedsUpdate(name, region string, requestedIP net.IP, ports []*api.ServicePort) (exists bool, needsUpdate bool, ipAddress string, err error) {
```
Can you add a comment indicating that requestedIP == nil is how you specify no requested IP.

cjcullen replied:
Also comment that the ipAddress returned is the IP address currently in use by the forwarding rule.

a-robinson replied:
Comment added.

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -452,42 +464,185 @@ func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, loadBalancerIP n
+	if len(fw.Allowed) != 1 {
+		return true, true, nil
+	}
+	if fw.Allowed[0].IPProtocol != "tcp" {
```
I think this would look better in the `if` above:
```
if len(fw.Allowed) != 1 || fw.Allowed[0].IPProtocol != "tcp"{
	return true, true, nil
}
```

a-robinson replied:
Sure, done

cjcullen commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -452,42 +464,185 @@ func (gce *GCECloud) EnsureTCPLoadBalancer(name, region string, loadBalancerIP n
+		Instances:       instances,
+		SessionAffinity: translateAffinityType(affinityType),
+	}
+	op, err := gce.service.TargetPools.Insert(gce.projectID, region, pool).Do()
```
We really need an `(op, error)` handling function :wink:

a-robinson replied:
Yeah, yeah, yeah... A little extra motivation is all I needed

a-robinson replied:
Actually, I'm not convinced it's all that useful, since I don't see any more than 2 functions that poll in the exact same way - some places have a regional op to poll while others have a global op; some consider StatusConflict a success, others consider StatusNotFound a success, while others consider them both failures; the delete functions give useful resource-specific wrappers around the errors.

I'm going to skip this for now.

thockin commented:
Just wanted to say thanks for this one - great change.  Thanks for jumping on review CJ

a-robinson committed:
Fix small comments.

a-robinson commented:
No need to re-review yet, I'm working on switching over to only using static IPs when updating a forwarding rule / target pool.

cjcullen commented:
+1 to only using static IPs during the update!

a-robinson commented:
Sigh, it makes handling firewalls correctly much trickier, though, since we either have to continue creating a static IP before creating the forwarding rule (even in the initial creation case) or we have to create the firewall after everything else is created, which breaks the nice invariant that calling GetTCPLoadBalancer returns success if and only if the load balancer is fully created.

a-robinson committed:
Only use static IPs for GCE forwarding rules while creating or updating
them, not in the steady state once they've been created. This makes it
much less likely that users will run into static IP quota issues.

Also add slightly more parallelism to the deletion of load balancers
now that I realize the static IPs can be deleted in parallel with
forwarding rules :)

a-robinson commented:
PTAL - running tests as we speak.

cjcullen commented:
LGTM.

cjcullen commented:
@k8s-bot test this

a-robinson commented:
Merging now to get additional soak time over the weekend. I'll revert later tonight if it looks like it's caused problems.
